pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -503,6 +509,11 @@ func (c *Controller) onNodeEvent(obj interface{}, event model.Event) error {
 		updatedNeeded = true
 		c.Lock()
 		delete(c.nodeInfoMap, node.Name)
+		// We should remove the node for all the services that
+		// had a workload there as we don't want to route traffic
+		for h := range c.serviceToWorkloadNodesMap {
+			delete(c.serviceToWorkloadNodesMap[h], node.Name)
```

Comment:
Concerned with the performance of the  loop

Reply:
This is when a node go down (which itself is not too frequent). The map contains only the gateway NodePort services whose external traffic policy is set to `Local`. I don't expect thousands of them to be present. So I thought a loop is fine. If you think we need to optimise this then it is possible to have an inverse map `node -> [nodePortWithLocalServices]` and only pick those keys while iterating.